enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,286 @@
+
+### Test Plan
+
+- Building the new ISO will be covered by the existing ZTP e2e tests.
+- A new ZTP job will need to be created for network booting.
+
+### Graduation Criteria
+
+This work will be GA immediately since there is no phasing possible or planned.
+Only ZTP will be affected.
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+This change does not introduce new resources, so upgrading or downgrading them
+is not a concern.
+
+For the new flow to apply, it must be explicitly requested through the
+`BareMetalHost` API. Otherwise, the old flow is used.
+
+There are no plans to remove the code that supports the *live ISO* flow from
+Ironic, Metal3 or OpenShift.
+
+### Version Skew Strategy
+
+The image-customization-controller will need to be updated before the new flow
+will be usable. Given that ZTP releases trail OpenShift releases (and are tied
+to them), it should not become an issue.
+
+### Operational Aspects of API Extensions
+
+#### Failure Modes
+
+- When image building fails in an `InfraEnv` resource, this failure will be
+  propagated to the baremetal-operator's `PreprovisioningImage` resource via
+  conditions. The whole provisioning process will be paused until the failure
+  is fixed.
+
+#### Support Procedures
+
+- The `image-customization-controller` logs will be the central place to
+  triage issues related to image building. Such issues will be highlighted
+  by a failure condition on the `PreprovisioningImage` resource.
+- The Ironic ramdisk logs (published via a special container
+  `ironic-ramdisk-logs`) will be useful to debug the coordination between
+  the two agents.
+
+## Implementation History
+
+- [Updated image building
+  process](https://github.com/openshift/image-customization-controller/pull/42)
+- [Ironic agent
+  plugin](https://github.com/openshift/ironic-agent-image/pull/40)
+
+## Drawbacks
+
+- The converged flow is more complex than each of the old flows.
+- The image-customization-controller becomes aware of (although does not
+  require) the assisted-service API.
+
+## Alternatives
+
+- Keep both flows separate, implement all required features (currently
+  firmware/BIOS settings and iPXE boot) twice.
+- The image-customization-controller could just ignore all CRs with an
+  `InfraEnv` link, and a separate controller (probably running in the same
+  process as BMAC) could reconcile those ones. A second place to vendor the
+  controller code just adds opportunities for it to get out of sync. The
+  image-customization-controller will still need to know about the `InfraEnv`
+  API anyway so it could ignore hosts with the label.
+- ICC may retrieve the Ignition from assisted-service and build the image
+  including the Ironic agent config itself, instead of the passing it to
+  assisted-service and having that assisted-image-service build the image.
```

Comment:
An advantage of this is that in future, a given InfraEnv could work both with hosts that are booted by Metal³ and those booted manually by the user.

In the current proposal, once any hosts in an InfraEnv have corresponding BMHs, i-c-c will update the InfraEnv ignition to contain IPA and not start AIA. This makes the ISO in the InfraEnv useless for a user who wants to manually boot it and use it as a discovery image.

Today this is not a problem, but it could limit us in future.

Reply:
cc @flaper87